pest/src/parser_state.rs

```diff
@@ -42,61 +48,151 @@ pub struct ParserState<'i, R: RuleType> {
     /// Specifies current atomicity
     pub atomicity: Atomicity,
     /// Stack of `Span`s
-    pub stack: Vec<Span<'i>>
+    pub stack: Stack<Span<'i>>
 }
 
 /// Creates a `ParserState` from a `&str`, supplying it to a closure `f`.
 ///
 /// # Examples
 ///
 /// ```
-/// # use std::rc::Rc;
 /// # use pest;
 ///
 /// let input = "";
-/// pest::state::<(), _>(input, |_, pos| {
-///     Ok(pos)
+/// pest::state::<(), _>(input, |s| {
+///     Ok(s)
 /// }).unwrap();
 /// ```
 pub fn state<'i, R: RuleType, F>(input: &'i str, f: F) -> Result<pairs::Pairs<'i, R>, Error<'i, R>>
 where
-    F: FnOnce(&mut ParserState<'i, R>, Position<'i>) -> Result<Position<'i>, Position<'i>>
+    F: FnOnce(Box<ParserState<'i, R>>) -> ParseResult<Box<ParserState<'i, R>>>
 {
-    let mut state = ParserState {
-        queue: vec![],
-        lookahead: Lookahead::None,
-        pos_attempts: vec![],
-        neg_attempts: vec![],
-        attempt_pos: 0,
-        atomicity: Atomicity::NonAtomic,
-        stack: vec![]
-    };
-
-    if f(&mut state, Position::from_start(input)).is_ok() {
-        let len = state.queue.len();
-        Ok(pairs::new(Rc::new(state.queue), input.as_bytes(), 0, len))
-    } else {
-        state.pos_attempts.sort();
-        state.pos_attempts.dedup();
-        state.neg_attempts.sort();
-        state.neg_attempts.dedup();
-
-        Err(Error::ParsingError {
-            positives: state.pos_attempts,
-            negatives: state.neg_attempts,
-            // All attempted positions were legal.
-            pos: unsafe { position::new(input.as_bytes(), state.attempt_pos) }
-        })
+    let state = ParserState::new(input);
+
+    match f(state) {
+        Ok(s) => {
```

Comment:
I would call this `state` instead of `s`, just to keep everything consistent, regardless of the shadowing.